Add support for Forgejo forge

pepicrft opened the pull request:
I'm user of [Codeberg](https://codeberg.org), a hosted instance of the Git forge [Forgejo](https://forgejo.org/), and I'd like to use UBI against repositories in those repos.

Since Forgejo mimics the API of GitHub, the implementation of Forgejo's forge symbol is very much a copy & paste of GitHub's. Please, let me know if you'd want me to extract some common functionality into another module for the purpose of reusability.

autarch commented:
Hi, thanks for your PR! I'm pretty finicky about my projects (see [this blog post](https://blog.urth.org/2022/07/02/what-s-the-right-way-to-merge-a-pull-request/) for details), so I rarely merge a PR as-is. I can move forward on your PR in one of two ways:

1. I check it out locally, fiddle with it as needed, merge it locally, and simply close this PR. This will preserve at least one commit with your name on it, but the PR will show up as closed in your GitHub stats.
2. If you enable me to push directly to your PR branch (which is the default when you make a PR), I can do my fiddling, then force push to your PR branch and merge the resulting PR. Again, this will preserve at least one commit with your name on it, but you _also_ get credit for the PR merge in your GitHub stats. The only downside is that I will be force pushing directly to your PR branch. **Note that this will not work if the PR branch is named `master`. GitHub doesn't allow me to push to the default branch of your fork.**

Please let me know which approach you'd prefer. If I don't hear from you before I get around to working on this PR I'll go with option 1.

Thanks again for your contribution!

autarch commented:
I spent some time trying to tweak this, but I finally realized that Forgejo does _not_ actually have the same API as GitHub, at least not the same as the GH v4 REST API.

Specifically, when you get releases from GitHub, you get something like this:

```json
{
  "assets": [
    {
      "url": "https://api.github.com/repos/houseabsolute/precious/releases/assets/237809592",
      "name": "precious-FreeBSD-x86_64.tar.gz",
  ],
}
```

But Forgejo doesn't include a `url` field for assets. You can _construct_ the URL from the asset's `id`, but even using that constructed URL, it doesn't seem like you can download the actual asset contents that way, unlike GitHub.

autarch commented:
All of which is to say, this _does_ need its own forge type. But don't work on that now. In working on this myself, I've done a bunch of refactoring that I'd like to finish up first.

pepicrft commented:
Hi @autarch  👋🏼 

First of all, thanks for your transparency. I'm open to whatever method works best for you (I'm not that attached to the idea of having to preserve my name in the work).

> But Forgejo doesn't include a url field for assets. You can construct the URL from the asset's id, but even using that constructed URL, it doesn't seem like you can download the actual asset contents that way, unlike GitHub.

This is something I missed completely. Is there anything I can do to help you understand this better?

Thanks a lot for helping move this one forward. I started moving some of my repositories to Codeberg, and being able to use UBI through Mise to install tools would be very valuable.

autarch commented:
So I think I'm done with the refactoring I was planning on doing. If you want to take a stab at adding support for ForgeJo, that'd be great!

pepicrft commented:
@autarch I doubt I'll find time in the following weeks to complete this work. Feel free to close it and I'll open a new one if I find the time for it.

autarch commented:
That's fine. I may take a look at some point, but right now I'm mostly filling my free time up by playing Satisfactory ;)

pepicrft commented:
@autarch I found some time to work on it. I'd appreciate if you could review the work and let me know how to proceed :)

autarch reviewed (changes requested):
This looks good. Thanks! There's a few things missing:

* Let's add `CODEBERG_TOKEN` as an env var to check for a token.
* The `builder.rs` code docs should mention the token env vars.
* The `ForgeType::from_url` code needs to accounts for the codeberg domain.
* It'd be great to add an integration test in `ubi-cli/tests/ubi.rs`. After fixing the above issue, I was able to get https://codeberg.org/Cyborus/forgejo-cli/releases installed using `ubi`, so I think this is a good one to use for the integration test. Note that it doesn't have a macOS release, so the test code needs to be configured to not run on macOS. You can see a bunch of examples of per-OS conditions in that file already.

Thanks again!

Finkregh commented on ubi/src/forge.rs:
```diff
@@ -50,9 +52,11 @@ pub(crate) trait Forge: std::fmt::Debug {
 
 const GITHUB_DOMAIN: &str = "github.com";
 const GITLAB_DOMAIN: &str = "gitlab.com";
+const FORGEJO_DOMAIN: &str = "codegerg.org";
```
Well, the software is forgejo, the default instance chosen here codeberg.org, so (apart from the typo) I think thats okay?

Finkregh commented on ubi/src/forge.rs:
```diff
@@ -50,9 +52,11 @@ pub(crate) trait Forge: std::fmt::Debug {
 
 const GITHUB_API_BASE: &str = "https://api.github.com";
 const GITLAB_API_BASE: &str = "https://gitlab.com/api/v4";
+const CODEBERG_API_BASE: &str = "https://codeberg.org/api/v1";
```
That should be `FORGEJO_API_BASE`, following the above variable?

Finkregh commented on ubi/src/forge.rs:
```diff
@@ -92,34 +96,41 @@ impl ForgeType {
             ForgeType::GitLab => Url::parse(GITLAB_API_BASE).unwrap(),
+            // The maintainers of Forgejo is Codeberg, hence why the URL
+            // doesn't align with the name of the Git forge.
+            ForgeType::Forgejo => Url::parse(CODEBERG_API_BASE).unwrap(),
```
But the api base changes depending on which forgejo instance you want to talk to?

pepicrft commented:
@Finkregh & @autarch thanks a lot for your review 🙏🏼 and apologies for the delay here. I addressed your comments and pushed the changes. Could you give it another look?

pepicrft commented:
> So given that, I think I want to refactor things a bit more to reduce code duplication. I think it's probably easiest for me to just take what you have and do the rest myself. Are you okay with that?

Totally :). I can't think of a better person to drive the refactors needed for that.

autarch commented:
This is in v0.8.0, which I just released. Thanks again!
